scipy/special/_logsumexp.py

```diff
@@ -93,43 +96,49 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     1.6094379124341005, 1.6094379124341005
 
     """
-    a = _asarray_validated(a, check_finite=False)
-    if b is not None:
-        a, b = np.broadcast_arrays(a, b)
-        if np.any(b == 0):
-            a = a + 0.  # promote to at least float
-            a[b == 0] = -np.inf
+    xp = array_namespace(a, b)
+    a, b = xp_broadcast_promote(a, b, ensure_writeable=True)
+    axis = tuple(range(a.ndim)) if axis is None else axis
+
+    if b is not None and xp.any(b==0):
+        a[b == 0] = -xp.inf
 
     # Scale by real part for complex inputs, because this affects
     # the magnitude of the exponential.
-    initial_value = -np.inf if np.size(a) == 0 else None
-    a_max = np.amax(a.real, axis=axis, keepdims=True, initial=initial_value)
+    if xp_size(a) == 0:
+        # because `xp.max` doesn't have `initial` argument...
+        shape = np.asarray(a.shape)  # NumPy is concise for scalar or tuple `axis`
+        shape[axis] = 1
+        a_max = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+    else:
+        real = xp.real(a) if xp.isdtype(a.dtype, "complex floating") else a
+        a_max = xp.max(real, axis=axis, keepdims=True)
 
     if a_max.ndim > 0:
-        a_max[~np.isfinite(a_max)] = 0
-    elif not np.isfinite(a_max):
+        a_max[~xp.isfinite(a_max)] = 0
```

Comment:
To support libraries with immutable arrays, you could write this as
```python
a_max = xp.where(xp.isfinite(a_max), a_max, 0)
```
although this will be less memory-efficient in the case of NumPy. Perhaps this is a good case for an array-compat utilty?

Reply:
Right, this has come up several times, as recently as yesterday - https://github.com/data-apis/array-api-compat/issues/144#issuecomment-2159444443. I think we need a function that mutates if possible and copies otherwise, but IIRC others wanted to wait for something before adding a utility.

Reply:
For now, we could also dispatch to JAX if it passes the tests.

Reply:
> wanted to wait for something

It's not clear to me what we're waiting for?

Reply:
cc @rgommers

Reply:
Let's bring this discussion back to https://github.com/scipy/scipy/pull/20085#issuecomment-2119336223.